numpy/core/src/multiarray/einsum.c.src

```diff
@@ -2874,12 +2862,16 @@ PyArray_EinsteinSum(char *subscripts, npy_intp nop,
 
         /* If the API was needed, it may have thrown an error */
         if (NpyIter_IterationNeedsAPI(iter) && PyErr_Occurred()) {
-            Py_DECREF(ret);
-            ret = NULL;
+            goto fail;
         }
     }
 
 finish:
+    if (out != NULL) {
+        ret = out;
```

Comment:
Out of curiousity, a couple of questions.

- The iterator iterates over `out` if present?
- The writeback to `out` or the new output array happens in iter deallocation?
- I assume deallocation also decrefs the `out` or new array?

Reply:
I need to incref the output before `NpyIterDeallocate` so it stays alive through that call.
- The assignment `op[nop] = out` near the beginning ensures `out`, if present is used in iteration, possibly with writeback semantics. Note that `nop` is the number of inputs, but actually `op` uses `nop +1` entries, one additional for the output.
- The writeback happens in `NpyIterDeallocate`, which also decrefs all operands including the output (which were increfed in `NpyIter_AdvancedNew`.
- In addition, `einsum` itself increfed all the inputs while deciding whether to create views or not, around `get_combined_dims_view`. So whether fail or finish, at the end the function decrefs the inputs, (not the output).